k2/python/csrc/torch/fsa_algo.cu

```diff
@@ -157,6 +158,22 @@ static void PybindArcSort(py::module &m) {
       py::arg("src"), py::arg("need_arc_map") = true);
 }
 
+static void PybindShortestPath(py::module &m) {
+  m.def(
+      "shortest_path",
+      [](Fsa &src, bool need_arc_indexes = true)
+          -> std::tuple<Fsa, torch::optional<torch::Tensor>, double> {
```

Comment:
I'd like to figure out how we'll deal with FsaVec before merging this.  The problem is, the input type is the same, but the return types would need to be different (i.e. FsaVec, Ragged<int32_t>, torch::Tensor).

Reply:
Wondering if we really need to support Fsa version, maybe it would be simpler if we just support version `FsaVec` as we may only need `FsaVec` version in production code? We can just wrap a funtion at Python level to convert Fsa to FsaVec with size 1 and do the simliar thing to the output (But I still doubt if we really need single Fsa version)

Reply:
How about adding `RaggedTensor` to Python which contains a C++ object `RaggedArray<T>` (for T == int32_t or float)?

```python
class RaggedTensor(object):
  def __init__(self, ragged_tensor: Union[_k2.RaggedIntArray, _k2.RaggedFloatArray]):
     pass
  def values() -> torch.Tensor:
     # return Ragged<T>::values
  def get(idx0: int) -> torch.Tensor:
    # if the ragged tensor has two axes, return the elements of `idx0`
```

We can return a `RaggedIntArray` to Python for a single FSA as well as a FsaVec.

Reply:
What I was thinking is, it would be easier in the long term if we created a RaggedTensor type in C++, which would be like `Ragged<X>` but the values would be a Tensor, and we expect its 1st axis to have dim == shape.NumElements().
Then the Python code can wrap that directly.
And we can define a template like this:
```
template <typename T>
Ragged<T> AsRagged(RaggedTensor &r);
```
which would expect the tensor to be convertible to that type and would crash if not.  Then when we need Ragged<X> in python we can use RaggedTensor, and convert in C++.

Reply:
Is the `Tensor` a `k2::Tensor` or a `torch::Tensor`.

Does the `RaggedTensor` in C++ look like the following:

```cpp
struct RaggedTensor {
  RaggedShape shape;
  Tensor values;  // REQUIRES: 1-D; continguous
};
```

> we expect its 1st axis to have dim == shape.NumElements().

Did you mean `TotSize(0)` or `TotSize(1)` ?

Reply:
The Tensor would be a k2::Tensor, because at this level we want the code to be independent of the outer toolkit.
By "we expect its 1st axis to have dim == shape.NumElements()", what I mean that the `tensor.Dim(0)` would equal the `shape.NumElements()`.